### ISSUE TITLE
Messages get same body as other message

### ISSUE THREAD

**Issue by mbilker** [nylas/N1#436]
Major bug for me. When testing my email-pgp plugin against multiple Facebook encrypted emails. I noticed it only decrypted the message for the first one, and used the cached version for the rest of the messages. It does not even hit my `MessageStoreExtension` to ask to format the other messages.

Here is an example:
![2015-11-19-224152_1280x800_scrot modified](https://cloud.githubusercontent.com/assets/1266541/11291558/d30be362-8f0f-11e5-9183-e3ce6f588df6.png)
vs.
![2015-11-19-224702_1280x800_scrot modified](https://cloud.githubusercontent.com/assets/1266541/11291551/c14ad6ec-8f0f-11e5-936f-ceafe41a3926.png)
which should be
![2015-11-19-230538_1280x800_scrot modified](https://cloud.githubusercontent.com/assets/1266541/11291772/6bc62e8a-8f12-11e5-9065-a39f566a023a.png)

**Comment by mbilker** [comment]
The issue is here: https://github.com/nylas/N1/blob/master/src/flux/stores/message-body-processor.coffee#L40-L41

It generates the key for the message before processing the message through the extensions. The body on the Facebook encrypted emails is `""` so the key for all the Facebook emails is the same and it doesn't reprocess the message.

**Comment by mbilker** [comment]
In the Nylas Developer Tools console, the MD5 hash of `""` results in
```
require('crypto').createHash('md5').update("").digest('hex')
"d41d8cd98f00b204e9800998ecf8427e"
```

**Comment by mbilker** [comment]
I don't know how to workaround the issue for the N1 project as a whole. Those message extensions can be computationally expensive.

I understand hashing the original body and caching the output of the extensions with the hash of the original body speeds up the process dramatically.

**Comment by mbilker** [comment]
I accidentally removed a block of code in mbilker/email-pgp#ba8c540 and readded it in mbilker/email-pgp#1a1048d

**Comment by bengotow** [comment]
Hi @mbilker — hmm, interesting... so the message body is identical in those messages, but the output /after/ the message body extensions run is unique?

If that's the case, I think we should change that hash function to use more than just the body to generate the hash. For example, we could say:

```
require('crypto').createHash('md5').update(message.id + message.body + message.subject).digest('hex')
```

Overall, the goal of that caching is to avoid re-processing the same message twice as messages are collapsed / uncollapsed, etc. and React renders are being triggered. It's alright if it runs for each message once.

**Comment by bengotow** [comment]
Hi @mbilker — I've updated the message body processor in 39a142ddcb80c7e1fce22dfe1e0e628272154523 so that the ID is part of the cache key. That way, every message will be processed, even if it has the same body as another message.

Also, I was talking with @grinich about the Facebook message decryption plugin you're working on. If you haven't already, it might be good to check out the "autoload-images" package (https://github.com/nylas/N1/tree/master/internal_packages/message-autoload-images).

It calls `MessageBodyProcessor.resetCache()` when the user changes their images whitelist so that it can re-process existing messages. You might want to do that when the encrypted data (in the email attachment) is finished downloading so you have a chance to re-process the message. In the future, we could add a `resetCacheForMessageId` method that would let you re-process for only a single message.

**Comment by mbilker** [comment]
@bengotow I actually found that before and really liked it! I noticed how it modifies the body of a message. Yes a `resetCacheForMessageId` would be nice, I currently implemented that sort of functionality at https://github.com/mbilker/email-pgp/blob/master/lib/message-loader/message-loader-header.es6#L47-L56

I do not plan on keeping the `email-pgp` plugin limited to just Facebook email. It would take some relearning of regular expressions to reliably detect PGP headers.
